cupy/cuda/memory.pyx

```diff
@@ -492,42 +493,80 @@ cdef class SingleDeviceMemoryPool:
                             raise OutOfMemoryError(size, total)
             chunk = Chunk(mem, 0, size)
 
-        chunk.in_use = True
-        self._in_use[chunk.ptr] = chunk
+        try:
+            rlock.lock_fastrlock(self._in_use_lock, -1, True)
+            self._in_use[chunk.ptr] = chunk
+        finally:
+            rlock.unlock_fastrlock(self._in_use_lock)
+        if remaining:
+            remaining_index = self._bin_index_from_size(remaining.size)
+            try:
+                rlock.lock_fastrlock(self._free_lock, -1, True)
+                self._free[remaining_index].add(remaining)
+            finally:
+                rlock.unlock_fastrlock(self._free_lock)
         pmem = PooledMemory(chunk, self._weakref)
         return MemoryPointer(pmem, 0)
 
     cpdef free(self, size_t ptr, Py_ssize_t size):
         cdef Chunk chunk
         cdef int index
 
-        chunk = self._in_use.pop(ptr, None)
+        try:
+            rlock.lock_fastrlock(self._in_use_lock, -1, True)
+            chunk = self._in_use.pop(ptr, None)
+        finally:
+            rlock.unlock_fastrlock(self._in_use_lock)
         if chunk is None:
             raise RuntimeError('Cannot free out-of-pool memory')
 
-        chunk.in_use = False
-        if chunk.next and not chunk.next.in_use:
+        if chunk.next:
+            chunk_next = None
             index = self._bin_index_from_size(chunk.next.size)
-            self._free[index].remove(chunk.next)
-            chunk = self._merge(chunk, chunk.next)
-
-        if chunk.prev and not chunk.prev.in_use:
+            try:
+                rlock.lock_fastrlock(self._free_lock, -1, True)
```

Comment:
Generally speaking, lock duration should be minimized as much as possible to minimize blocking duration.
I even moved `chunk = self._merge(chunk, chunk_next)` to outside of the lock to minimize blocking duration.
I understand that you cared of overhead of calling `lock` and `unlock`.
However, I am pretty sure that minimized locking overcomes larger locking in terms of performance with multi-threaded application.